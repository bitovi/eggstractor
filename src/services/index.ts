export * from './collection.service';
export * from './token.service';
export * from './variable.service'; 