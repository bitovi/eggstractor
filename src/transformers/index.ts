export * from "./scss.transformer";
export * from "./css.transformer";
