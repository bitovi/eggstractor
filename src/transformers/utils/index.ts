export * from './deduplicate-messages.utils';
export * from './get-style-property-and-value.utils';
export * from './group-by.utils';
export * from './update-padding-based-on-border.utils';
