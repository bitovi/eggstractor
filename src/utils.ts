export default {
  toBase64(str: string): string {
    const base64chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    const utf8str = encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
      (_, p1) => String.fromCharCode(parseInt(p1, 16))
    );
    let i = 0;
    let result = '';
    while (i < utf8str.length) {
      const char1 = utf8str.charCodeAt(i++);
      const char2 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const char3 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const enc1 = char1 >> 2;
      const enc2 = ((char1 & 3) << 4) | (char2 >> 4);
      const enc3 = ((char2 & 15) << 2) | (char3 >> 6);
      const enc4 = char3 & 63;
      result += base64chars[enc1] + base64chars[enc2] +
        (isNaN(char2) ? '=' : base64chars[enc3]) +
        (isNaN(char3) ? '=' : base64chars[enc4]);
    }
    return result;
  },
  sanitizeName(name: string): string {
    return name.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  sanitizeSegment(segment: string): string {
    return segment.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  rgbToHex(r: number, g: number, b: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    return '#' + rHex + gHex + bHex;
  },
  rgbaToString(r: number, g: number, b: number, a: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    const aHex = Math.round(a * 255).toString(16).padStart(2, '0');
    
    return a === 1 ? 
      '#' + rHex + gHex + bHex : 
      '#' + rHex + gHex + bHex + aHex;
  },
  groupBy<T>(arr: T[], key: (item: T) => string): Record<string, T[]> {
    return arr.reduce((groups, item) => {
      const k = key(item);
      if (!groups[k]) groups[k] = [];
      groups[k].push(item);
      return groups;
    }, {} as Record<string, T[]>);
  }
};