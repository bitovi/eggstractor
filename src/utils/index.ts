export * from './color.utils';
export * from './create-naming-convention'
export * from './constants.utils';
export * from './delay.utils';
export * from './get-valid-stylesheet-format.utils';
export * from './node.utils';
export * from './string.utils';
export * from './test.utils';
export * from './units.utils';
export * from './value.utils';
