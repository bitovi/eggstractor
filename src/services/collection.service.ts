import { ComponentSetToken, ComponentToken, TokenCollection } from '../types';
import { getProcessorsForNode } from '../processors';
import { extractComponentSetToken, extractComponentToken, extractNodeToken } from '../services';
import { getNodePathNames } from '../utils/node.utils';

function getFlattenedValidNodes(node: BaseNode): BaseNode[] {
  const result: BaseNode[] = [];

  function traverse(currentNode: BaseNode) {
    // Skip VECTOR and INSTANCE nodes entirely. INSTANCE nodes which are
    // instances of components, which are not relevant for token extraction.
    if ('type' in currentNode && ['VECTOR', 'INSTANCE'].includes(currentNode.type)) {
      return;
    }

    result.push(currentNode);

    if ('children' in currentNode) {
      for (const child of currentNode.children) {
        traverse(child);
      }
    }
  }

  traverse(node);
  return result;
}

export async function collectTokens(onProgress: (progress: number, message: string) => void) {
  const collection: TokenCollection = { tokens: [], components: {}, componentSets: {} };

  let componentToken: ComponentToken | null = null;
  let componentSetToken: ComponentSetToken | null = null;

  let totalNodes = 0;
  let processedNodes = 0;

  let lastTimestamp = Date.now();
  let lastPercentage = -1;

  async function processNode(node: BaseNode) {
    processedNodes++;

    // calculate integer percent in your 10–95 range
    const currentPercentage = Math.floor((processedNodes / totalNodes) * 85) + 10;
    const shouldUpdate = currentPercentage !== lastPercentage || processedNodes === totalNodes;

    if (shouldUpdate) {
      lastTimestamp = currentPercentage;
      onProgress(currentPercentage, `Processing nodes… ${processedNodes}/${totalNodes}`);

      // throttle yields to at most once every 200 ms
      const now = Date.now();
      if (now - lastTimestamp >= 200) {
        await new Promise((r) => setTimeout(r, 0));
        lastTimestamp = now;
      }
    }

    if ('type' in node && 'boundVariables' in node) {
      if (node.type === 'COMPONENT_SET') {
        componentSetToken = extractComponentSetToken(node);
        collection.componentSets[node.id] = componentSetToken;
      }

      if (node.type === 'COMPONENT') {
        componentToken = extractComponentToken(node, componentSetToken!);
        collection.components[node.id] = componentToken;
      }

      const nodePathNames = getNodePathNames(node);
      const processors = getProcessorsForNode(node);

      for (const processor of processors) {
        const tokens = await extractNodeToken(
          node,
          processor,
          nodePathNames,
          componentToken,
          componentSetToken,
        );
        collection.tokens.push(...tokens);
      }
    }
  }

  onProgress(0, 'Loading pages...');
  await figma.loadAllPagesAsync();

  onProgress(5, 'Counting nodes...');
  const validNodes = figma.root.children.flatMap((page) => getFlattenedValidNodes(page));
  totalNodes = validNodes.length;

  onProgress(10, `Processing ${totalNodes} nodes...`);

  // Process all pages in parallel for maximum speed
  const pagePromises = figma.root.children.map(async (page) => {
    const validNodes = getFlattenedValidNodes(page);

    for (const node of validNodes) {
      await processNode(node);
    }
  });

  await Promise.all(pagePromises);

  return collection as Readonly<TokenCollection>;
}
