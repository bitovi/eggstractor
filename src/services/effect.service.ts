import { VariableToken, TokenCollection } from '../types';
import { sanitizeName } from '../utils';

/**
 * Convert a Figma EffectStyle to a VariableToken
 */
function createEffectStyleToken(effectStyle: EffectStyle): VariableToken | null {
  try {
    // Convert effects array to CSS box-shadow value
    const boxShadowValues = effectStyle.effects
      .filter((effect) => effect.type === 'DROP_SHADOW' || effect.type === 'INNER_SHADOW')
      .map((effect) => {
        const { offset, radius, color, spread } = effect;
        const isInset = effect.type === 'INNER_SHADOW';

        // Convert RGBA to CSS rgba()
        const colorValue = `rgba(${Math.round(color.r * 255)}, ${Math.round(color.g * 255)}, ${Math.round(color.b * 255)}, ${color.a ?? 1})`;

        // Build box-shadow value: [inset] offset-x offset-y blur-radius [spread-radius] color
        const parts = [];
        if (isInset) parts.push('inset');
        parts.push(`${offset?.x || 0}px`);
        parts.push(`${offset?.y || 0}px`);
        parts.push(`${radius || 0}px`);
        if (spread) parts.push(`${spread}px`);
        parts.push(colorValue);

        return parts.join(' ');
      });

    if (boxShadowValues.length === 0) {
      return null;
    }

    const rawValue = boxShadowValues.join(', ');

    return {
      type: 'variable',
      path: [{ name: effectStyle.name, type: 'FRAME' }],
      property: 'box-shadow',
      name: sanitizeName(effectStyle.name),
      value: `$${sanitizeName(effectStyle.name)}`,
      rawValue,
      valueType: null,
      metadata: {
        figmaId: effectStyle.id,
        variableName: effectStyle.name,
        variableTokenType: 'primitive',
      },
    };
  } catch (error) {
    console.warn(`Failed to create token for effect style ${effectStyle.name}:`, error);
    return null;
  }
}

/**
 * Collect all Figma Effect Styles (box-shadow, etc.)
 */
export async function collectAllFigmaEffectStyles(
  collection: TokenCollection,
  onProgress: (progress: number, message: string) => void,
) {
  try {
    onProgress(7, 'Collecting Figma Effect Styles...');
    const effectTokens: VariableToken[] = [];

    // Get all local effect styles
    const effectStyles = await figma.getLocalEffectStylesAsync();

    for (const effectStyle of effectStyles) {
      onProgress(8, `Processing effect style: ${effectStyle.name}`);

      const token = createEffectStyleToken(effectStyle);
      if (token) {
        effectTokens.push(token);
      }
    }

    console.log(`âœ¨ Collected ${effectTokens.length} effect style tokens`);
    collection.tokens.push(...effectTokens);
  } catch (error) {
    console.warn('Failed to collect Figma Effect Styles:', error);
  }
}
