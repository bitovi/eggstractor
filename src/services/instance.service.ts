import { StyleToken, InstanceToken, ComponentToken, TokenCollection } from '../types';
import { getProcessorsForNode } from '../processors';
import { extractNodeToken } from './token.service';
import { getNodePathNames } from '../utils/node.utils';

export interface InstanceOverride {
  instanceId: string;
  instanceName: string;
  componentId?: string;
  overrideTokens: StyleToken[];
  allTokens: StyleToken[];
}

/**
 * Detects and extracts override styles for component instances by comparing
 * fresh style extractions from both instance and component nodes.
 */
export async function detectInstanceOverrides(
  collection: TokenCollection
): Promise<InstanceOverride[]> {
  const overrides: InstanceOverride[] = [];

  for (const [instanceId, instanceToken] of Object.entries(collection.instances)) {
    if (instanceToken.remote || !instanceToken.componentNode) {
      continue; // Skip remote instances and instances without component references
    }

    try {
      // Get the actual Figma nodes for comparison
      const instanceNode = await figma.getNodeByIdAsync(instanceId) as InstanceNode;
      const componentNode = instanceToken.componentNode;
      
      if (!instanceNode || !componentNode) {
        continue;
      }

      // Extract fresh styles from both nodes for accurate comparison
      const componentSetToken = instanceToken.componentNode?.parent?.type === 'COMPONENT_SET' 
        ? collection.componentSets[instanceToken.componentNode.parent.id]
        : undefined;
      
      const componentTokenData = collection.components[componentNode.id];
      
      const instanceStyles = await extractInstanceStyles(instanceNode, componentTokenData, componentSetToken);
      const componentStyles = await extractComponentStyles(componentNode, componentTokenData, componentSetToken);

      // Find overrides by comparing properties
      const overrideTokens: StyleToken[] = [];
      
      for (const instanceStyle of instanceStyles) {
        const correspondingComponentStyle = componentStyles.find(
          componentStyle => componentStyle.property === instanceStyle.property
        );

        // If there's no corresponding token in the component, or values differ, it's an override
        if (!correspondingComponentStyle || 
            correspondingComponentStyle.value !== instanceStyle.value ||
            correspondingComponentStyle.rawValue !== instanceStyle.rawValue) {
          overrideTokens.push(instanceStyle);
        }
      }

      // Include this instance even if there are no overrides (as per requirement)
      overrides.push({
        instanceId,
        instanceName: instanceToken.name,
        componentId: componentNode.id,
        overrideTokens,
        allTokens: instanceStyles,
      });

    } catch (error) {
      console.warn(`Failed to process instance ${instanceToken.name}:`, error);
      // Still include the instance with empty overrides
      overrides.push({
        instanceId,
        instanceName: instanceToken.name,
        componentId: instanceToken.componentNode?.id,
        overrideTokens: [],
        allTokens: [],
      });
    }
  }

  return overrides;
}

/**
 * Extracts all style properties for an instance node by processing it directly.
 * This is used to get fresh token data for comparison.
 */
export async function extractInstanceStyles(
  instanceNode: InstanceNode,
  componentToken?: ComponentToken,
  componentSetToken?: any
): Promise<StyleToken[]> {
  const nodePathNames = getNodePathNames(instanceNode);
  const processors = getProcessorsForNode(instanceNode);
  const allTokens: StyleToken[] = [];

  for (const processor of processors) {
    const tokens = await extractNodeToken(
      instanceNode,
      processor,
      nodePathNames,
      componentToken,
      componentSetToken,
    );

    const styleTokens = tokens.filter((token): token is StyleToken => token.type === 'style');
    allTokens.push(...styleTokens);
  }

  return allTokens;
}

/**
 * Extracts all style properties for a component node by processing it directly.
 */
export async function extractComponentStyles(
  componentNode: ComponentNode,
  componentToken?: ComponentToken,
  componentSetToken?: any
): Promise<StyleToken[]> {
  const nodePathNames = getNodePathNames(componentNode);
  const processors = getProcessorsForNode(componentNode);
  const allTokens: StyleToken[] = [];

  for (const processor of processors) {
    const tokens = await extractNodeToken(
      componentNode,
      processor,
      nodePathNames,
      componentToken,
      componentSetToken,
    );

    const styleTokens = tokens.filter((token): token is StyleToken => token.type === 'style');
    allTokens.push(...styleTokens);
  }

  return allTokens;
}
