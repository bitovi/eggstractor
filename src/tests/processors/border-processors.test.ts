import { collectTokens } from '../../services';
import { transformToScss } from '../../transformers';
import { createTestData } from '../../utils';
import testData from '../fixtures/figma-test-data_border-position.json';
import testDataSides from '../fixtures/figma-test-data_border-sides.json';
import testDataShape from '../fixtures/figma-test-data_border-shape.json';
import testDataRadius from '../fixtures/figma-test-data_border-radius.json';

describe('Border Processors', () => {
  it('should process border correctly', async () => {
    const { setupTest } = createTestData(testData);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(jest.fn());
    const { result } = transformToScss(tokens);

    expect(result).toMatchSnapshot('border');
  });

  it('should process border sides correctly', async () => {
    const { setupTest } = createTestData(testDataSides);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(jest.fn());
    const { result } = transformToScss(tokens);

    expect(result).toMatchSnapshot('border-sides');
  });

  it('should process border shape correctly', async () => {
    const { setupTest } = createTestData(testDataShape);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(jest.fn());
    const { result } = transformToScss(tokens);

    expect(result).toMatchSnapshot('border-shape');
  });

  it('should process border radius correctly', async () => {
    const { setupTest } = createTestData(testDataRadius);
    const testSetup = await setupTest();

    global.figma = testSetup.figma;

    const tokens = await collectTokens(jest.fn());
    const { result } = transformToScss(tokens);

    expect(result).toMatchSnapshot('border-radius');
  });
});
