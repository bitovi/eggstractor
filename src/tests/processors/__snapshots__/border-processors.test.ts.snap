// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Border Processors should process border correctly: border 1`] = `
"// Generated SCSS Variables
$border-size-primary: 0.25rem;
$border-primary: #687879;

// Generated SCSS Mixins
@mixin position-inside-custom {
  box-shadow: inset 0 $border-size-primary 0 0 $border-primary, inset (-$border-size-primary) 0 0 0 $border-primary, inset 0 (-$border-size-primary) 0 0 $border-primary, inset $border-size-primary 0 0 0 $border-primary;
}
@mixin position-center-custom {
  border: $border-size-primary solid $border-primary;
}
@mixin position-outside-custom {
  outline: $border-size-primary solid $border-primary;
}
"
`;

exports[`Border Processors should process border correctly: border 2`] = `
"// Generated SCSS Variables
$border-size-primary: 0.25rem;
$border-primary: #687879;

// Generated SCSS Mixins
@mixin position-inside-custom {
  box-shadow: inset 0 $border-size-primary 0 0 $border-primary, inset (-$border-size-primary) 0 0 0 $border-primary, inset 0 (-$border-size-primary) 0 0 $border-primary, inset $border-size-primary 0 0 0 $border-primary;
}
@mixin position-center-custom {
  border: $border-size-primary solid $border-primary;
}
@mixin position-outside-custom {
  outline: $border-size-primary solid $border-primary;
}
"
`;

exports[`Border Processors should process border radius correctly: border-radius 1`] = `
"// Generated SCSS Variables
$colors-primary--brand--teal-teal-800: #00464a;
$corner-radius-primary: 1rem;
$size-null: 0rem;

// Generated SCSS Mixins
@mixin border-radius-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $corner-radius-primary;
}
@mixin border-radius-topright-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $size-null $corner-radius-primary $size-null $size-null;
}
@mixin border-radius-bottomright-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $size-null $size-null $corner-radius-primary;
}
@mixin border-radius-topleft-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $corner-radius-primary $size-null $size-null;
}
@mixin border-radius-bottomleft-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $size-null $size-null $size-null $corner-radius-primary;
}
@mixin border-radius-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 1rem;
}
@mixin border-radius-topright-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 0 1rem 0 0;
}
@mixin border-radius-bottomright-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 0 0 1rem;
}
@mixin border-radius-topleft-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 1rem 0 0;
}
@mixin border-radius-bottomleft-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 0 0 0 1rem;
}
"
`;

exports[`Border Processors should process border radius correctly: border-radius 2`] = `
"// Generated SCSS Variables
$colors-primary--brand--teal-teal-800: #00464a;
$corner-radius-primary: 1rem;
$size-null: 0rem;

// Generated SCSS Mixins
@mixin border-radius-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $corner-radius-primary;
}
@mixin border-radius-topright-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $size-null $corner-radius-primary $size-null $size-null;
}
@mixin border-radius-bottomright-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $size-null $size-null $corner-radius-primary;
}
@mixin border-radius-topleft-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $corner-radius-primary $size-null $size-null;
}
@mixin border-radius-bottomleft-variable {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: $size-null $size-null $size-null $corner-radius-primary;
}
@mixin border-radius-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 1rem;
}
@mixin border-radius-topright-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 0 1rem 0 0;
}
@mixin border-radius-bottomright-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 0 0 1rem;
}
@mixin border-radius-topleft-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 1rem 0 0;
}
@mixin border-radius-bottomleft-custom {
  background: $colors-primary--brand--teal-teal-800;
  border-radius: 0 0 0 1rem;
}
"
`;

exports[`Border Processors should process border shape correctly: border-shape 1`] = `
"// Generated SCSS Variables
$border-size-primary: 0.25rem;
$border-primary: #687879;

// Generated SCSS Mixins
@mixin shape-rectangle {
  box-shadow: inset 0 $border-size-primary 0 0 $border-primary, inset (-$border-size-primary) 0 0 0 $border-primary, inset 0 (-$border-size-primary) 0 0 $border-primary, inset $border-size-primary 0 0 0 $border-primary;
}
@mixin shape-rounded-rectangle {
  box-shadow: inset 0 $border-size-primary 0 0 $border-primary, inset (-$border-size-primary) 0 0 0 $border-primary, inset 0 (-$border-size-primary) 0 0 $border-primary, inset $border-size-primary 0 0 0 $border-primary;
  border-radius: 1.25rem;
}
@mixin shape-ellipse {
  border: $border-size-primary solid $border-primary;
  box-shadow: inset 0 0.25rem 0 0 $border-primary, inset -0.25rem 0 0 0 $border-primary, inset 0 -0.25rem 0 0 $border-primary, inset 0.25rem 0 0 0 $border-primary;
  border-radius: 50%;
}
@mixin shape-line-curved {
  border: $border-size-primary solid $border-primary;
}
"
`;

exports[`Border Processors should process border shape correctly: border-shape 2`] = `
"// Generated SCSS Variables
$border-size-primary: 0.25rem;
$border-primary: #687879;

// Generated SCSS Mixins
@mixin shape-rectangle {
  box-shadow: inset 0 $border-size-primary 0 0 $border-primary, inset (-$border-size-primary) 0 0 0 $border-primary, inset 0 (-$border-size-primary) 0 0 $border-primary, inset $border-size-primary 0 0 0 $border-primary;
}
@mixin shape-rounded-rectangle {
  box-shadow: inset 0 $border-size-primary 0 0 $border-primary, inset (-$border-size-primary) 0 0 0 $border-primary, inset 0 (-$border-size-primary) 0 0 $border-primary, inset $border-size-primary 0 0 0 $border-primary;
  border-radius: 1.25rem;
}
@mixin shape-ellipse {
  border: $border-size-primary solid $border-primary;
  box-shadow: inset 0 0.25rem 0 0 $border-primary, inset -0.25rem 0 0 0 $border-primary, inset 0 -0.25rem 0 0 $border-primary, inset 0.25rem 0 0 0 $border-primary;
  border-radius: 50%;
}
@mixin shape-line-curved {
  border: $border-size-primary solid $border-primary;
}
"
`;

exports[`Border Processors should process border sides correctly: border-sides 1`] = `
"// Generated SCSS Variables
$border-size-primary: 0.25rem;
$size-null: 0rem;
$border-primary: #687879;
$size-16: 1rem;

// Generated SCSS Mixins
@mixin border-center-solid-custom-topleft {
  border-top: 0.125rem solid #869597;
  border-left: 0.125rem solid #869597;
}
@mixin border-center-solid-custom-top {
  border: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-topleft {
  border-top: $border-size-primary solid $border-primary;
  border-left: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-variable-top {
  border: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-custom-left {
  border-right: 0.5rem solid #869597;
  border-left: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-left {
  border-right: $size-16 solid $border-primary;
  border-left: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-custom-right {
  border-right: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-right {
  border-right: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-custom-bottom {
  border-bottom: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-bottom {
  border-bottom: $border-size-primary solid $border-primary;
}
"
`;

exports[`Border Processors should process border sides correctly: border-sides 2`] = `
"// Generated SCSS Variables
$border-size-primary: 0.25rem;
$size-null: 0rem;
$border-primary: #687879;
$size-16: 1rem;

// Generated SCSS Mixins
@mixin border-center-solid-custom-topleft {
  border-top: 0.125rem solid #869597;
  border-left: 0.125rem solid #869597;
}
@mixin border-center-solid-custom-top {
  border: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-topleft {
  border-top: $border-size-primary solid $border-primary;
  border-left: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-variable-top {
  border: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-custom-left {
  border-right: 0.5rem solid #869597;
  border-left: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-left {
  border-right: $size-16 solid $border-primary;
  border-left: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-custom-right {
  border-right: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-right {
  border-right: $border-size-primary solid $border-primary;
}
@mixin border-center-solid-custom-bottom {
  border-bottom: 0.125rem solid #869597;
}
@mixin border-center-solid-variable-bottom {
  border-bottom: $border-size-primary solid $border-primary;
}
"
`;
